pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/SchedulerManager.java

```diff
@@ -215,28 +239,147 @@ public synchronized void initialize(Producer<byte[]> exclusiveProducer) {
+            }
+        } else {
+            throw new DrainInProgressException();
+        }
+    }
+
+    public LongRunningProcessStatus getDrainStatus(String workerId) {
         long startTime = System.nanoTime();
+        String errString;
+        LongRunningProcessStatus retVal = new LongRunningProcessStatus();
+        try {
+            val workerStatus = drainOpStatusMap.get(workerId);
+            if (workerStatus == null) {
+                errString = "Worker " + workerId + " not found in drain records";
+                retVal = LongRunningProcessStatus.forError(errString);
+            } else {
+                switch (workerStatus) {
+                    default:
+                        errString = "getDrainStatus: Unexpected status " + workerStatus + " found for worker " + workerId;
+                        retVal = LongRunningProcessStatus.forError(errString);
+                        break;
+                    case DrainCompleted:
+                        retVal = LongRunningProcessStatus.forStatus(LongRunningProcessStatus.Status.SUCCESS);
+                        break;
+                    case DrainInProgress:
+                        retVal = LongRunningProcessStatus.forStatus(LongRunningProcessStatus.Status.RUNNING);
+                        break;
+                    case DrainNotInProgress:
+                        retVal = LongRunningProcessStatus.forStatus(LongRunningProcessStatus.Status.NOT_RUN);
+                        break;
+                }
+            }
+        } finally {
+            log.info("Get drain status for worker {} - execution time: {} sec; returning status={}, error={}",
+                    workerId, (System.nanoTime() - startTime) / Math.pow(10, 9), retVal.status, retVal.lastError);
+            return retVal;
+        }
+    }
+
+    @VisibleForTesting
+    void clearDrainOpsStatus() {
+        drainOpStatusMap.clear();
+        log.warn("Cleared drain op status map");
+    }
+
+    @VisibleForTesting
+    void setDrainOpsStatus(final String workerId, final DrainOpStatus dStatus) {
+        drainOpStatusMap.put(workerId, dStatus);
+        log.warn("setDrainOpsStatus: updated drain status of worker {} to {}", workerId, dStatus);
+    }
 
+    @VisibleForTesting
+    ConcurrentHashMap<String, DrainOpStatus> getDrainOpsStatusMap() {
+        val retVal = new ConcurrentHashMap<String, DrainOpStatus>(drainOpStatusMap);
+        return retVal;
+    }
+
+    private synchronized int getCurrentAvailableNumWorkers() {
+        return getCurrentAvailableWorkers().size();
+    }
+
+    private synchronized Set <String> getCurrentAvailableWorkers() {
         Set<String> currentMembership = membershipManager.getCurrentMembership()
                 .stream().map(workerInfo -> workerInfo.getWorkerId()).collect(Collectors.toSet());
+        val unavailableWorkers = new ArrayList<String>();
+        // iterate the set, instead of the concurrent hashmap
+        for (String worker : currentMembership) {
```

Comment:
We don't need two loops to remove.  You can use an iterator to remove with iterating the set.

https://mkyong.com/java/java-how-to-remove-items-from-a-list-while-iterating/